자동완성 추가기능 구현

HannaJeon opened the pull request:
- prototype -> class로 변경
- webpack 빌드 적용
- 자동완성 최근검색결과 기능 추가

HannaJeon committed:
prototype -> class로 변경

HannaJeon committed:
const와 let으로 변수 변경

HannaJeon committed:
ES6 template 적용

HannaJeon committed:
Promise패턴 적용

HannaJeon committed:
polyfill코드를 polyfill.js로 이동

HannaJeon committed:
polyfill 파일 삭제

kkiluk committed:
빌드 환경을 구축하다.

webpack을 이용한 빌드 환경을 구축했습니다.
- 수동 빌드: npm run build
- 실시간 빌드: npm run watch

kkiluk committed:
최근 검색어 리스트의 화면을 추가하다.

최근 검색어 리스트를 보여줄 recent-keyword-list를 추가했습니다.

kkiluk committed:
자동완성 데이터 localStorage에 저장하다.

기존의 memo, memoLog를 localStorage의 acData, acDataLog로 저장했습니다.

HannaJeon committed:
디렉토리 최상단에 위치했던 app.js 삭제

HannaJeon committed:
디렉토리 최상단에 위치했던 app.js 삭제

HannaJeon committed:
자동완성 데이터 캐시하기 미션 코드 작성하다.

checkValidation 메소드로 cache시간과 데이터정보 확인하여 true일 경우 cache데이터에서 정보 확인.
false일 경우 ajax통신

HannaJeon committed:
recentKeywordList 생성

kkiluk committed:
최근검색어 관리 로직을 구현하다.

돋보기 버튼을 클릭하여 검색을 수행할 때, 검색어를 localStorage에 저장했습니다.
최근검색어는 최대 5개까지 관리합니다.

HannaJeon committed:
최근검색결과 삭제 버튼 생성

HannaJeon committed:
최근검색결과 html 적용

HannaJeon committed:
Merge branch 'caching' into teamB-ES6

HannaJeon committed:
webpack을 이용한 테스트환경 구축

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+        this.setLocalStorageItem('acDataLog', this.acDataLog);
+    }
+    cacheRecentData(key) {
+        if(!key) {
```
이정도면 아예 cacheRecentData를 부르기전에 체크하는게 더 나을 듯.

HannaJeon replied:
완료

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+        this.setLocalStorageItem('recentData', this.recentData);
+    }
+    checkValidation(keyword) {
+        const expiredTime = 6 * 60 * 60 * 1000
```
최종결과값을 그냥 적어두어도 좋겠네요. 호출될때마다 계산하게 하지 말구요.

HannaJeon replied:
완료

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+    checkValidation(keyword) {
+        const expiredTime = 6 * 60 * 60 * 1000
+        // const expiredTime =  5 * 1000
+        if(this.acData.hasOwnProperty(keyword) && (Date.now() - this.acData[keyword].create) < expiredTime) {
```
if문안에 내용이 좀 긴데요 좀 줄여서 쓰죠. 
Date.now() - this.acData[keyword].create 이정도 코드는 변수로 저장해두고 쓴다던가.
그런 습관이 디버깅을 쉽게해줍니다. (빨리 변수값을 확인할 수가 있어서)

HannaJeon replied:
완료

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+
+        this.domContainer.searchField.addEventListener('keydown', this.checkKeyCode.bind(this));
+        this.domContainer.searchField.addEventListener('input', this.changeSearchText.bind(this));
+        this.domContainer.searchField.addEventListener('focusin', this.clickSearchField.bind(this));
```
DOM을 반복적으로 탐색하게 하지말고, 변수로 담가두고 쓰세요.
var searchField =  this.domContainer.searchField;
searchField.addEventlistener()//

HannaJeon replied:
완료

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+        this.acRenderer.updateRecentList(this.acResource.recentData)
+    }
+    checkKeyCode(e) {
+		switch(e.keyCode){
```
switch문을 없앨수도 있는데요.
아래처럼 객체를 만들어두면, 

```javascript
const KeyboardMap = {
  38 : this.acRenderer.pressUpKey,
  39 : this.acRenderer.pressDownKey,
  40 : this.acRenderer. pressEnterKey
}

//이렇게 호출가능 하겠죠.
KeyboardMap[e.keyCode]();
```

HannaJeon replied:
완료

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+		}
+	}
+    changeSearchText(e) {
+		const keyword = e.target.value;
```
별건 아니지만 들여쓰기가 들쭉날쭉이라 github에 올라간 코드가 보기 어렵네요.

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+            return;
+        }
+        const parent = target.parentNode.parentNode;
+        const index = Array.prototype.indexOf.call(parent.children, target.parentNode);
```
Array.from 메서드도 있어요.

HannaJeon replied:
Array.from으로 작성하기로 합의 후 변경

kkiluk committed:
자동완성 테스트코드 리뷰를 반영하다.

ACResponder 생성자의 파라미터를 하나의 오브젝트로 묶었습니다.
currHoveredItem.previousElementSibling와 currHoveredItem.nextElementSibling를 지역변수로 저장하여 dom cache했습니다.

crongro commented on bundle.js:
```diff
@@ -0,0 +1,332 @@
+        if(!keyword) {
+            this.domContainer.recentKeywordList.style.display = "block";
+        } else {
+            this.domContainer.recentKeywordList.style.display = "none";
```
전체코드에서 element를 캐시해서 쓰는 습관이 필요해요.

crongro reviewed (approved):
리뷰 참고하세요.
